README.md
# CNTT

[![GitHub Release Date](https://img.shields.io/github/release-date/t3n0/cntt)](https://github.com/t3n0/cntt/releases/latest)
[![GitHub release (latest by date)](https://img.shields.io/github/v/release/t3n0/cntt)](https://github.com/t3n0/cntt/releases/latest)
[![GitHub all releases](https://img.shields.io/github/downloads/t3n0/cntt/total)](https://github.com/t3n0/cntt/releases/download/v0.4/cntt-v0.4.zip)
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)

![logo-wide](./docs/figures/logo-wide.png)

**CNTT**, also spelled **C-entity**, is an utility to compute, display and manipulate electronic and optical properties of single-walled carbon nanotubes (SWCNTs). CNTT is an acronym for **Carbon NanoTube Tool**.

At the current state of developement, this tool is capable of computing the electronic and excitonic band structure of SWCNTs, along with the electron and exciton Density of States (DOS). This is done in a very efficient way by exploiting the **helical symmetry** of SWCNTs. Results are then displayed in helical coordinates. In this way, most of the optical properties of SWCNTs become apparent by simply inspecting the [helical bandstructure](./docs/figures/cnt(4,2).png).

## Installation

Simply download [.zip file](https://github.com/t3n0/cntt/releases/download/v0.4/cntt-v0.4.zip), extract it at your favourite location and run

```Shell
pip install .
```

This will install the system-wide command `cntt` and the python package `cntt`.

### DFT capabilities

Since version v0.3, Density Function Theory (DFT) capabilities are included in CNTT.
These are outsourced to [Quantum Espresso](https://www.quantum-espresso.org/) which must be available in the local PATH.
To download and install QE please refer to the [official QE documentation](https://www.quantum-espresso.org/Doc/user_guide_PDF/user_guide.pdf).

## Usage (1): command line

For a basic usage of the tool, simply type in the terminal

```Shell
cntt 4 2
```

This will display:

- the linear and helical band structure of a [(4,2) single-walled carbon nanotube](./docs/figures/cnt(4,2).png),
- the (singlet, bright and dark) exciton energy dispersions,
- electron and exciton DOS,
- and the most important physical parameters, unit cells and Brillouin zones.

Also, typing `cntt -h` provide a help dialoge for advanced usage.

## Usage (2): package

From a pyhton interpreter, import the base `Swcnt` class. For a basic usage, just copy the following snippet:

```Python
from cntt.swcnt import Swcnt

mycnt = Swcnt(4,2)

mycnt.calculateCuttingLines()

# tight-binding calculation with onsite energy 3.0eV and Fermi level at 0.0eV
mycnt.calculateElectronBands('TB', 'TB',  sym = 'lin', gamma = 3.0, fermi = 0.0)
mycnt.calculateElectronBands('TB', 'TB0', sym = 'hel', gamma = 3.0, fermi = 0.0)

# same as above, but with different fermi level
mycnt.calculateElectronBands('TB', 'TB1', sym = 'hel', gamma = 3.0, fermi = 1.5)

mycnt.calculateKpointValleys()

# excitons generated by the two different underlying electron structures
mycnt.calculateExcitonBands('EM','TB0', deltaK = 10, bindEnergy = 0.2)
mycnt.calculateExcitonBands('EM','TB1', deltaK = 10, bindEnergy = 0.1)

mycnt.calculateDOS('electron')
mycnt.calculateDOS('exciton')

mycnt.plot()
```

The above code produces the following plot

![cnt(4,2)fermi](./docs/figures/cnt(4,2)fermi.png)

## Examples and tutorials

CNTT provides a number of functionalities to compute, manipulate and visualise SWCNTs properties.
The following links are simple jupyter notebooks that explain the core capabilities of CNTT:

- [basic usage](./docs/tutorials/1.basics.ipynb): zone-filding TB, linear and helical BZs.
- [simple dft](./docs/tutorials/2.dft-coarse.ipynb): coarse k-grid DFT, comparison with the TB model.
- [finer dft](./docs/tutorials/3.dft_restart.ipynb): restart a previous DFT calculation on a finer k-grid.

## Support

For any problems, questions or suggestions, please contact me at [tenobaldi@gmail.com](tenobaldi@gmail.com).

## Roadmap

Currently, this project provides:

- plotting the unit cells in three different configurations (the cnt supercell N-atom flake, the linear 2-atom cell and the helical 2-atom cell);
- visualizing the corresponding graphene-like Brillouin zones;
- computing and displaying the carbon nanotube band structure from the tight-binding zone-folding approximation;
- DFT calculation of electronic bands (outsourced by [Quantum Espresso](https://www.quantum-espresso.org/));
- computing and displaying the dispersion relation of bright and dark singlet excitons;
- computing and displaying the DOS of electron and excitons;
- simultaneous band plotting for comparison;
- basic doping capabilities (fermi level and binding energy can be tuned and will influence the exciton dispersion);

Future developments will include:

- joint density of states (JDOS);
- improved tight-binding methods;
- phonon calculations;
- output of the 3D carbon atoms coordinates;
- exciton absorption spectrum;
- optical matrix elements;

## Authors and acknowledgment

The development of CNTT is proudly powered by [me](https://github.com/t3n0).
Also, please consider citing the relevant literature if you are going to use this tool:

- [Carbon 186, 465-474 (2022)](https://doi.org/10.1016/j.carbon.2021.10.048)

The DFT capabilities of CNTT are provided by [Quantum Espresso](https://www.quantum-espresso.org/manifesto/).
QE is licensed under GPL3 and citation guidelines can be found [here](https://www.quantum-espresso.org/Doc/user_guide/node6.html#SubSec:Terms) and [here](https://www.quantum-espresso.org/Doc/user_guide/node3.html).

## License

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program.  If not, see <https://www.gnu.org/licenses/>.


CHANGELOG
cntt 0.4 (2023 ?? ??)
---------------------------

* Changed
  - The CNT constructor now takes the ksteps and symmetry parameters.
  - The <calculateCuttingLines> method has been replaced with the <setKSteps> method.
  - Labelling bands has been removed. CNT objects now only contain one set of bands.

* Added
  - CHANGELOG file.
  - Exciton 2D maps, so to visualize exciton in the (ke, kh) space.
  - Electron-hole continuum bands.

* Improved
  - Exciton band dispersion now span the full helical BZ.
  - Plotting direct and reciprocal lattice now can handle multiple cnts at once.

* Bugfix
  - Bug in the calculation of the helical reciprocal vectors: beta was wrong due to integer casting.


cntt 0.3 (2023 Mar 29)
---------------------------

* Added
  - DFT capabilities (outsourced to Quantum Espresso).
  - Fourier interpolation of k-points.

* Bugfix
  - Minor fixes.


cntt 0.2 (2022 Oct 28)
---------------------------

* Added
  - Package has been renamed to CNTT.
  - New logo.
  - New plotting features: labels, simultaneous multiple plots and comparison.
  - Density of States calculation (for both electron and excitons).
  - Basic doping capabilities: the choice of the fermi level and binding energy will reflect in the exciton band dispersion.

* Improved
  - Plot of the kpoint valley minima (takes into account valence and conduction bands).

* Bugfix
  - Minor fixes.


swcnt-bands 0.1 (2022 Sep 20)
---------------------------

* Added
  - Linear and helical unit cells.
  - Linear and helical Brillouin zones.
  - Linear and helical electron band structure.
  - Helical effective mass approximation of the dispersion relation of bright and dark singlet excitons.


cntt/swcnt.py
'''
CNTT / C-entity
-----------

CNTT, also spelled C-entity, is a software for computing, displaying and
manipulating the physical properties of single-walled carbon nanotubes (SWCNTs).

CONTACTS:
---------

email: tenobaldi@gmail.com
github: https://github.com/t3n0/cntt

LICENSE:
--------

Copyright (c) 2021-2023 Stefano Dal Forno.

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
'''


import os
import numpy as np
import numpy.linalg as la

import cntt.physics as physics
import cntt.tightbinding as tightbinding
import cntt.dft as dft


class Swcnt(object):
    """
    Base class for computing the physical properties
    of single-walled carbon nanotubes (SWCNTs).
    Properties include:
        - direct and reciprocal lattice structure;
        - electron and exciton energy dispersions;
        - density of states (DOS) and joint density of states (JDOS);
        - absorption spectra (soon);
        - and more (soon).
    """
    def __init__(self, n, m, ksteps = 100, a0 = 0.2461, sym = 'helical'):
        '''
        Constructor for the Swcnt class.

        Parameters:
        -----------
            n: int
                carbon nanotube n number

            m: int
                carbon nanotube m number

            ksteps: int (optional)
                number of kpoints for the BZ discretisation
                default = 100

            a0: float (optional)
                carbon nanotube lattice constant (nm)
                default = 0.2461 nm
            
            sym: str (optional)
                symmetry of the BZ, either linear or helical
                default = 'helical'
        '''

        # units
        self.unitL = 'nm'
        self.unitE = 'eV'
        self.unitInvL = 'nm-1'

        # graphene constants
        self.a0 = a0 # nm
        self.b0 = 4 * np.pi / np.sqrt(3) / self.a0 # nm-1
        self.ac = self.a0/np.sqrt(3)
        self.bc = self.b0/np.sqrt(3)
        self.cellVolume = self.a0**2 * np.sqrt(3) / 2
        self.bzVolume   = self.b0**2 * np.sqrt(3) / 2

        # carbon nanotube parameters
        self.n, self.m = n, m
        self.R = np.gcd(2 * m + n, 2 * n + m)
        self.D = np.gcd(m, n)
        self.N = 2 * (n ** 2 + n * m + m ** 2) // self.R
        self.t1, self.t2 = (2 * m + n) // self.R, -(2 * n + m) // self.R

        # graphene lattice vectors
        self.a1 = self.a0 * np.array([np.sqrt(3) / 2, 1 / 2])
        self.a2 = self.a0 * np.array([np.sqrt(3) / 2, -1 / 2])
        self.b1 = self.b0 * np.array([1 / 2, np.sqrt(3) / 2])
        self.b2 = self.b0 * np.array([1 / 2, -np.sqrt(3) / 2])

        # CNT Supercell lattice vectors (C, T)
        self.C = self.n * self.a1 + self.m * self.a2
        self.T = self.t1 * self.a1 + self.t2 * self.a2

        # Linear 2 atom unit cell (Tl, T)
        self.l1, self.l2 = physics.twoAtomUnitCell(self.t1, self.t2, self.n)
        self.Tl = self.l1 * self.a1 + self.l2 * self.a2
        alpha = self.n * self.l2 - self.m * self.l1

        # Helical 2 atom unit cell (Th, C/D)
        self.h1, self.h2 = physics.twoAtomUnitCell(self.n // self.D, self.m // self.D, self.t1)
        self.Th = self.h1 * self.a1 + self.h2 * self.a2
        beta = self.h2 * self.t1 - self.h1 * self.t2

        # CNT reciprocal lattice vectors
        self.K1 = (-self.t2 * self.b1 + self.t1 * self.b2) / self.N  # K1.T = 0
        self.K2 = (self.m * self.b1 - self.n * self.b2) / self.N     # K2.C = 0

        self.k1L = self.N * self.K1
        self.k2L = self.K2 + self.K1 * alpha
        
        self.k1H = self.K1 * self.D + self.K2 * beta
        self.k2H = - self.K2 * self.N / self.D

        # CNT norms
        self.normT = la.norm(self.T)
        self.normC = la.norm(self.C)
        self.normK1 = la.norm(self.K1)
        self.normK2 = la.norm(self.K2)
        self.bzLin = la.norm(self.K2)
        self.bzHel = la.norm(self.k2H)
        self.normOrt = self.bzLin * abs(beta) / self.D

        # Symmetry and cuttiling lines
        self.kSteps = ksteps
        if sym in ['linear', 'lin', 'l', 'L']:
            self.sym = 'linear'
            self.cuttingLines = physics.bzCuts(self.K2, self.K1, self.N, self.kSteps)
            self.kGrid = np.linspace(0.0, 1.0, self.kSteps, endpoint=False) * self.bzLin
            self.subN = self.N
        else:
            self.sym = 'helical'
            self.cuttingLines = physics.bzCuts(self.k2H, self.k1H / self.D, self.D, self.kSteps)
            self.kGrid = np.linspace(0.0, 1.0, self.kSteps, endpoint=False) * self.bzHel
            self.subN = self.D

        # CNT data containers for electron and exciton bands, DOS, JDOS, etc
        self.electronBands    = np.empty((self.subN), dtype=object)
        self.electronDOS      = np.empty((self.subN), dtype=object)

        self.excitonBands     = np.empty((self.subN, self.subN), dtype=object)
        self.excitonContinuum = np.empty((self.subN, self.subN), dtype=object)
        self.excitonMaps      = np.empty((self.subN, self.subN), dtype=object)
        self.excitonDOS       = np.empty((self.subN, self.subN), dtype=object)
        

    def setUnits(self, energy, length):
        '''
        Set the energy and length unit of measure.
        This is just a wrapper for the output.
        The code performs all calculations in eV and nm.

        Parameters:
        -----------
            energy: str
                energy units [ 'eV' | 'Ha' | 'Ry' ]

            length: str
                length units [ 'nm' | 'bohr' | 'Angstrom' ]
        '''
        self.unitE = energy
        self.unitL = length
        self.unitInvL = length + '-1'


    def setKSteps(self, ksteps):
        '''
        Reset the numer of kpoints for the BZ discretisation.

        (Useful to increase the ksteps after a coarse calculation
        and subsequently to interpolate with Fourier)

        Parameters:
        -----------
            ksteps: int (optional)
                number of kpoints for the discretisation
        '''
        self.kSteps = ksteps
        if self.sym == 'linear':
            self.cuttingLines = physics.bzCuts(self.K2, self.K1, self.N, self.kSteps)
            self.kGrid = np.linspace(0.0, 1.0, self.kSteps, endpoint=False) * self.bzLin
            self.subN = self.N
        else:
            self.sym = 'helical'
            self.cuttingLines = physics.bzCuts(self.k2H, self.k1H / self.D, self.D, self.kSteps)
            self.kGrid = np.linspace(0.0, 1.0, self.kSteps, endpoint=False) * self.bzHel
            self.subN = self.D


    def calculateElectronBands(self, calc, **kwargs):
        '''
        Calculate the CNT electron band structure using different methods.

        Parameters:
        -----------
            calc: str
                specify the method for the band calculation
                [ 'TB' | 'DFT' ]

            **kwargs: (optional)
                Key-value arguments that depend on the calculation to be performed.

                calc = 'TB' calculation:
                ------------------------
                    gamma: float (optional)
                        TB on-site parameter
                        default = 3.0 eV
                    
                    fermi: float (optional)
                        position of the Fermi energy wrt the graphene Fermi level
                        default = 0.0 eV
                
                calc = 'DFT' calculation:
                -------------------------
                    nprocs: int (optional)
                        number of processors to use
                        default = 1
                    
                    from_file: bool (optional)
                        if True, read bands from a previous calculation
                        default = False

                    fourier_interp: bool (optional)
                        if True, the code will attempt  to interpolate the dft bands
                        to the BZ kpoint grid
                        default = False

                    pseudo_dir: str (optional)
                        directory containing the pseudopotential
                        default = './pseudo_dir'
                    
                    ecutwfc: float (optional)
                        wavefunction cutoff for the planewave expansion
                        default = 20 Ry
                    
                    ecutrho: float (optional)
                        density cutoff for the planewave expansion
                        default = 200 Ry
                    
                    nbnd: int (optional)
                        number of electron bands to calculate
                        default = 8
                    
                    deltan: int (optional)
                        number of bands around the Fermi energy to consider
                        default = 1

                    clat: float (optional)
                        spacing between adjecent graphene sheets
                        default = 1 nm
                    
                    kpoints: int (optional)
                        number of kpoints for the BZ sampling in the self-consistent calculation
                        default = 12
        '''
        if calc == 'TB':
            tightBindingElectronBands(self, **kwargs)
        elif calc == 'DFT':
            dft.dftElectronBands(self, **kwargs)
        elif calc == 'something else':
            pass
        else:
            print(f'Calculation {calc} not implemented.')


    # def calculateKpointValleys(self):
    #     '''
    #     Calculate the (kx,ky) position of the minima and maxima of the electron band dispersion.
    #     Note! The K valley extrema are calculated in the helical coordinate system.
    #     '''
    #     bzCuts = self.bzCutsHel
    #     bands = self.electronBands
    #     subN, _, _, _ = bands.shape

    #     self.condKpointValleys = []
    #     self.valeKpointValleys = []

    #     for mu in range(0, subN): # angular momentum
    #         vales, conds = physics.valeCondBands(bands[mu])
    #         condMasks = mathematics.findFunctionListExtrema(conds, 'min')
    #         valeMasks = mathematics.findFunctionListExtrema(vales, 'max')
    #         cuts = bzCuts[mu]
            
    #         # valence band properties
    #         kValley = []
    #         for mask in valeMasks:
    #             kValley.append( cuts[mask] )
    #         self.valeKpointValleys.append( kValley )
            
    #         # conduction band properties
    #         kValley = []
    #         for mask in condMasks:
    #             kValley.append( cuts[mask] )
    #         self.condKpointValleys.append( kValley )


    def calculateExcitonBands(self, calc, **kwargs):
        '''
        Calculate the exciton bands energy dispersion using different methods.
        
        Note! Excitons are always computed in helical coordinates. This means
        that the underlying electron bands should also have helical symmetry.

        Parameters:
        -----------
            calc: str
                specify the method for the band calculation
                [ 'free' | 'BSE' | 'more in the future' ]

            **kwargs: (optional)
                key-value arguments depend on the calculation to be performed.

                calc = 'free' calculation:
                ---------------------------                    
                    bindEnergy: float (optional)
                        binding energy between electron and hole states
                        default = 0.0 eV
        '''
        if calc == 'free':
            freeExcitonBands(self, **kwargs)
        elif calc == 'BSE':
            pass
        elif calc == 'more in the future':
            print('Pollo!')
        else:
            print(f'Calculation {calc} not implemented.')


    def calculateDOS(self, which, eMin=None, eMax=None, eSteps=1000):
        '''
        Calcualte the density of states for a given particle energy dispersion.
        By default, the DOS is calculated on the helical symmetry.

        Parameters:
        -----------
            which: str
                particle type [ 'electron' | 'exciton' ]

            eMin: float (optional)
                minimum of the DOS energy grid
                default, minimum of the band
            
            eMax: float (optional)
                maximum of the DOS energy grid
                default, maximum of the band

            eSteps: int (optional)
                energy steps for the discretisation
                default = 1000
        '''
        if which in ['electron', 'elec', 'el']:
            self.electronEnergyGrid = physics.energyGrid(eMin, eMax, eSteps, self.electronBands.tolist())
            for mu in range(self.subN):
                dos = []
                for band in self.electronBands[mu]:
                    dos.append( physics.densityOfStates(band, self.kGrid, self.electronEnergyGrid) )
                self.electronDOS[mu] = dos
        elif which in ['exciton', 'exc', 'ex']:
            self.excitonEnergyGrid = physics.energyGrid(eMin, eMax, eSteps, self.excitonBands.tolist())
            for mu1 in range(self.subN):
                for mu2 in range(self.subN):
                    dos = []
                    for band in self.excitonBands[mu1, mu2]:
                        dos.append( physics.densityOfStates(band, self.kGrid, self.excitonEnergyGrid) )
                    self.excitonDOS[mu1, mu2] = dos
        else:
            print(f'Particle {which} not recognized.')
        


    def calculateJDOS(self, which, enSteps=1000):
        '''
        Calcualte the joint density of states for a given particle energy dispersion.
        By default, the JDOS is calculated on the helical symmetry.

        Parameters:
        -----------
            which: str
                particle type [ 'el' | 'ex' | 'ph' ]

            enesteps: int (optional)
                energy steps for the discretisation
                default = 1000
        '''
        pass


    def getCell(self, which):
        '''
        Returns a polygon (rectangle) representing the selected cell.

        Parameters:
        -----------
            which: str
                Selection of the CNT cell to return:
                supercell, unit cell, reciprocal cell
                ['sc' | 'uc' | 'rc']
        
        Returns:
        --------
            res: ndarray
                array of (x, y) positions of the four corners of the selected cell
        '''
        if which in ['supercell', 'sc']:
            res = [ [0.0, 0.0], self.T, self.T + self.C, self.C]
        elif which in ['unitcell', 'uc']:
            if self.sym == 'linear':
                res = [ [0.0, 0.0], self.Tl, self.Tl + self.T, self.T]
            else:
                res = [ [0.0, 0.0], self.C / self.D, self.C / self.D + self.Th, self.Th ]
        elif which in ['reciprocalcell', 'rc']:
            if self.sym == 'linear':
                res = [ [0.0, 0.0], self.k1L, self.k1L + self.k2L, self.k2L ]
            else:
                res = [[0.0, 0.0], self.k1H, self.k1H + self.k2H, self.k2H]
        return np.array(res)


    def saveToDirectory(self, dirpath):
        WORK_DIR = os.getcwd()
        OUT_DIR = os.path.join(WORK_DIR, dirpath)
        if not os.path.exists(OUT_DIR):
            os.makedirs(OUT_DIR)
        
        print(OUT_DIR)
        # for mu in range(0, self.NU):
        #     path = os.path.join(dirpath, f"bandLin{mu:03d}.txt")
        #     utils.save_file(self.bandLin[mu][0], self.bandLin[mu][1], path=path)
        # for mu in range(0, self.D):
        #     path = os.path.join(dirpath, f"bandHel{mu:03d}.txt")
        #     utils.save_file(self.bandHel[mu][0], self.bandHel[mu][1], path=path)
        # for k in self.excParaBands:
        #     path = os.path.join(dirpath, f"excPara{k}.txt")
        #     utils.save_file(*self.excParaBands[k], path=path)
        # for k in self.excPerpBands:
        #     path = os.path.join(dirpath, f"excPerp{k}.txt")
        #     utils.save_file(*self.excPerpBands[k], path=path)
        # for k in self.excDarkBands:
        #     path = os.path.join(dirpath, f"excDark{k}.txt")
        #     utils.save_file(*self.excDarkBands[k], path=path)

        # path = os.path.join(dirpath, f"cicco.txt")
        # utils.save_file(np.array([1,2,3]), np.array([4,5,6]), path=path)

        excitons = self.excitonBands['DFT-65rough']
        
        print(len(excitons))
        for exc in excitons:
            # path = os.path.join(dirpath, f"{exc}.txt")
            # utils.save_file(*excitons[exc], path=path)
            #np.savetxt(path, excitons[exc])
            #print(exc, ' ', excitons[exc])
            print(exc)


    # def plot(self, path=None):
    #     '''
    #     Plots all relevant quantities and properties of the nanotube object.

    #     This is a *convenience* function that displays the carbon nanotube
    #     parameters, direct and reciprocal lattices, K and K' valley minima,
    #     linear and helical electron bands, and the helical exciton and
    #     phonon bands.

    #     For more versatile plotting routines, the user might want to import
    #     the `cntt.plotting` module.

    #     Parameters:
    #     -----------
    #         path: str (optional)
    #             destination path where to save the figure
    #             default = shows the figure
    #     '''
    #     if path == None:
    #         fig = plt.figure(figsize=(16, 9))
    #     else:
    #         fig = plt.figure(figsize=(16, 9), dpi=300)

    #     # figure and axes
    #     fig.suptitle(f"CNT ({self.n},{self.m})")
    #     ax1 = fig.add_axes([0.23, 0.63, 0.35, 0.32])
    #     ax2 = fig.add_axes([0.63, 0.63, 0.35, 0.32])
    #     ax3 = fig.add_axes([0.05, 0.25, 0.2, 0.3])
    #     ax4 = fig.add_axes([0.25, 0.25, 0.58, 0.3])
    #     ax5 = fig.add_axes([0.83, 0.25, 0.15, 0.3])
    #     ax6 = fig.add_axes([0.25, 0.05, 0.58, 0.2])
    #     ax7 = fig.add_axes([0.83, 0.05, 0.15, 0.2])

    #     # plot all axes
    #     plotting.dirLat(self, ax1)
    #     plotting.recLat(self, ax2)
    #     plotting.electronBands(self, ax3, 'lin')
    #     plotting.electronBands(self, ax4, 'hel')
    #     plotting.excitonBands(self, ax6)
    #     plotting.electronDOS(self, ax5, True)
    #     plotting.excitonDOS(self, ax7, True)

    #     # format some text and layout
    #     ax3.set_title("Linear")
    #     ax4.set_title("Helical")
    #     ax5.set_title("DOS")
    #     for ax in [ax4, ax5, ax7]:
    #         ax.set_ylabel('')
    #     for ax in [ax4, ax5]:
    #         ax.set_xlabel('')

    #     # "Matplotlib fai ribrezzo ai popoli del mondo!"
    #     # "Tu metti in subbuglio il mio sistema di donna sensibile!"
    #     # (Alida Valli)

    #     ax3.get_shared_y_axes().join(ax3, ax4, ax5)
    #     ax6.get_shared_y_axes().join(ax6, ax7)                                   
    #     ax4.get_shared_x_axes().join(ax4, ax6)
        
    #     ax4.set_yticklabels([]) 
    #     ax5.set_yticklabels([])
    #     ax4.set_xticklabels([])
    #     ax5.set_xticklabels([])
    #     ax7.set_yticklabels([])

    #     # cnt parameters text
    #     plt.text(0.05, 0.9, utils.textParams(self), ha="left", va="top", transform=fig.transFigure)

    #     # ticks locators and grids
    #     for ax in [ax4, ax6]:
    #         ax.locator_params(axis='x', nbins=10)
    #     for ax in [ax3, ax4, ax5]:
    #         ax.locator_params(axis='y', nbins=6)
    #     for ax in [ax6, ax7]:
    #         ax.locator_params(axis='y', nbins=6)
    #     for ax in [ax3, ax4, ax5, ax6, ax7]:    
    #         ax.grid(linestyle='--')


    #     # save if path is not None
    #     if path == None:
    #         plt.show()
    #     else:
    #         fig.savefig(path)
    #         plt.close()


def tightBindingElectronBands(cnt: Swcnt, gamma=3.0, fermi=0.0):
    '''
    Computes the band structure of the given CNT with the tight binding method in the zone folding scheme.
    Bands are computed using the symmetry selected in the <cuttingLines> method.
    '''
    for mu, cut in enumerate(cnt.cuttingLines):
        upperBand =   tightbinding.grapheneTBBands(cut, cnt.a1, cnt.a2, gamma) - fermi
        lowerBand = - tightbinding.grapheneTBBands(cut, cnt.a1, cnt.a2, gamma) - fermi
        cnt.electronBands[mu] = np.array( [lowerBand, upperBand] )
    


def freeExcitonBands(cnt: Swcnt, bindEnergy=0.0):
    '''
    Computes the excitonic band structure in helical coordinates.
    The electron-hole Coulomb interaction is not taken into account.
    The bands are computed in a non-interacting approximation, hence
    a constant value for the binding energy can be input by the user
    (default = 0 eV).
    '''
    for mu1 in range(cnt.subN):
        bandLo = cnt.electronBands[mu1][0,:]
        
        for mu2 in range(cnt.subN):    
            bandUp = cnt.electronBands[mu2][1,:]
            
            bands, continuum, map = physics.excBand1Band2(bandLo, bandUp, cnt.kGrid)
            cnt.excitonBands[mu1, mu2] = bands - bindEnergy
            cnt.excitonContinuum[mu1, mu2] = continuum
            cnt.excitonMaps[mu1, mu2] = map



